tests/Authenticator.hs

```diff
@@ -0,0 +1,426 @@
+{-# LANGUAGE BinaryLiterals #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Authenticator
+  ( PublicKeyCredentialSource (..),
+    Authenticator (..),
+    authenticatorMakeCredential,
+    authenticatorGetAssertion,
+  )
+where
+
+import qualified Client.PrivateKey as PrivateKey
+import qualified Codec.CBOR.Write as CBOR
+import Control.Monad (forM_, unless, when)
+import qualified Crypto.Fido2.Model as M
+import qualified Crypto.Fido2.Operations.Attestation.None as None
+import qualified Crypto.Fido2.PublicKey as PublicKey
+import Crypto.Hash (hash)
+import qualified Crypto.PubKey.ECC.Generate as ECC
+import qualified Crypto.PubKey.ECC.Types as ECC
+import qualified Crypto.PubKey.Ed25519 as Ed25519
+import Crypto.Random (MonadRandom)
+import qualified Crypto.Random as Random
+import Data.Binary.Put (runPut)
+import qualified Data.Binary.Put as Put
+import qualified Data.ByteArray as BA
+import qualified Data.ByteString as BS
+import Data.ByteString.Lazy (toStrict)
+import Data.List (find)
+import qualified Data.Map as Map
+import Data.Maybe (fromMaybe, mapMaybe)
+import qualified Data.Set as Set
+import Data.Text.Encoding (encodeUtf8)
+
+-- | [(spec)](https://www.w3.org/TR/webauthn-2/#public-key-credential-source)
+-- A stored credential.
+data PublicKeyCredentialSource = PublicKeyCredentialSource
```

Comment:
It is possible to remedy this by adding one additional map to the authenticator; ``Map CredentialId SignatureCounter``. And removing the publicKey altogether. We need the publicKey for signing, but we could also derive the PublicKey from the PrivateKey when needed.

I propose leaving the signaturecounter where it is, but removing the publicKey, and documenting that we have a per-credential counter that we store as part of the source.

Reply:
Removing the `publicKey` is pretty easy:
```diff
diff --git a/tests/Authenticator.hs b/tests/Authenticator.hs
index f7296a1..c9b3287 100644
--- a/tests/Authenticator.hs
+++ b/tests/Authenticator.hs
@@ -40,7 +40,6 @@ data PublicKeyCredentialSource = PublicKeyCredentialSource
   { pkcsCounter :: M.SignatureCounter,
     pkcsId :: M.CredentialId,
     pkcsPrivateKey :: PrivateKey.PrivateKey,
-    pkcsPublicKey :: PublicKey.PublicKey,
     pkcsRpId :: M.RpId,
     pkcsUserHandle :: Maybe M.UserHandle
   }
@@ -199,7 +198,6 @@ authenticatorMakeCredential
               { pkcsCounter = M.SignatureCounter 0,
                 pkcsId = credentialId,
                 pkcsPrivateKey = privateKey,
-                pkcsPublicKey = publicKey,
                 pkcsRpId = rpId,
                 pkcsUserHandle = Just userHandle
               }
@@ -399,7 +397,6 @@ authenticatorGetAssertion
       -- a variable length) is always the last element.
       signature <-
         PrivateKey.sign
-          (pkcsPublicKey selectedCredential)
           (pkcsPrivateKey selectedCredential)
           (M.adRawData authenticatorData <> BA.convert (M.unClientDataHash clientDataHash))
 
diff --git a/tests/Client/PrivateKey.hs b/tests/Client/PrivateKey.hs
index f190dfd..11e2517 100644
--- a/tests/Client/PrivateKey.hs
+++ b/tests/Client/PrivateKey.hs
@@ -37,12 +37,11 @@ toECDSAKey PublicKey.COSEAlgorithmIdentifierES384 = pure . ES384PrivateKey
 toECDSAKey PublicKey.COSEAlgorithmIdentifierES512 = pure . ES512PrivateKey
 toECDSAKey _ = const $ fail "Not a ECDSA key identifier"
 
-sign :: MonadRandom m => PublicKey.PublicKey -> PrivateKey -> ByteString -> m Signature
-sign (PublicKey.ES256PublicKey _) (ES256PrivateKey key) msg = ESSignature <$> ECDSA.sign key Hash.SHA256 msg
-sign (PublicKey.ES384PublicKey _) (ES384PrivateKey key) msg = ESSignature <$> ECDSA.sign key Hash.SHA384 msg
-sign (PublicKey.ES512PublicKey _) (ES512PrivateKey key) msg = ESSignature <$> ECDSA.sign key Hash.SHA512 msg
-sign (PublicKey.Ed25519PublicKey publicKey) (Ed25519PrivateKey privateKey) msg = pure . Ed25519Signature $ Ed25519.sign privateKey publicKey msg
-sign _ _ _ = error "Unknown or incorrect private/public keypair"
+sign :: MonadRandom m => PrivateKey -> ByteString -> m Signature
+sign (ES256PrivateKey key) msg = ESSignature <$> ECDSA.sign key Hash.SHA256 msg
+sign (ES384PrivateKey key) msg = ESSignature <$> ECDSA.sign key Hash.SHA384 msg
+sign (ES512PrivateKey key) msg = ESSignature <$> ECDSA.sign key Hash.SHA512 msg
+sign (Ed25519PrivateKey privateKey) msg = pure . Ed25519Signature $ Ed25519.sign privateKey (Ed25519.toPublic privateKey) msg
 
 toByteString :: Signature -> ByteString
 toByteString (ESSignature ECDSA.Signature {..}) = ASN1.encodeASN1' ASN1.DER [ASN1.Start ASN1.Sequence, ASN1.IntVal sign_r, ASN1.IntVal sign_s, ASN1.End ASN1.Sequence]
```

Reply:
Regarding signature counter, it might also be pretty nice to move it outside to an authenticator field, as it would allow us to encompass different supported signature counter methods, something like this:
```haskell
data SignatureCounter
  = Unsupported           
  | Global Word32
  | PerCredential (Map CredentialId Word32)
```

Reply:
Agreed! I will implement these!